### PR TITLE
Fix smtp sender return error on close

### PR DESCRIPTION
**Pull request by rikysya**
What: Fix smtp sender returns an error because of incorrect resources closing

Why: 

Thanks for submitting a PR!

## Code Review Checklist (to be filled out by reviewer)
 - [ ] Does the PR describe what changes are being made?
 - [ ] Does the PR describe why the changes are being made?
 - [ ] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [ ] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [ ] Could the PR be broken into smaller PRs?
 - [ ] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [ ] Does any documentation need updating?

### CODE REVIEW

File: internal/post/sender.go [hunk 1]
```diff
@@ -45,8 +46,11 @@ func (sender *SMTPSender) SendEmail(ctx context.Context, msg *Message) (err erro
 		return err
 	}
 	// close underlying connection
+	// if any unexpected error occurred
 	defer func() {
-		err = errs.Combine(err, client.Close())
+		if err != nil {
```
Comment (egonelbre) [inline internal/post/sender.go]: Currently when an error happens earlier then the connection won't be closed.

File: internal/post/sender.go [hunk 2]
```diff
@@ -73,26 +77,31 @@ func (sender *SMTPSender) SendEmail(ctx context.Context, msg *Message) (err erro
 		}
 	}
 
-	data, err := client.Data()
+	mess, err := msg.Bytes()
```
Comment (egonelbre) [inline internal/post/sender.go]: Why was this moved upward?

File: internal/post/sender.go (hunk 1 above)
Comment (egonelbre) [reply internal/post/sender.go]: It feels like it always should call `client.Close`, but depending on the outer error might ignore it.

Comment (rikysya) [reply internal/post/sender.go]: @egonelbre  if there is no error that means we called `client.Quit` successfully, which closes the connection. If error some error occurs during sending process then we close it manually in defer

File: internal/post/sender.go (hunk 2 above)
Comment (rikysya) [reply internal/post/sender.go]: for `data` to be closer to function that invokes it

File: internal/post/sender.go (hunk 1 above)
Comment (rikysya) [reply internal/post/sender.go]: could you please describe a situation when it won't be closed?

Comment (egonelbre) [reply internal/post/sender.go]: Let's say msg.Bytes() returns a parsing error.

Comment (rikysya) [reply internal/post/sender.go]: @egonelbre so then we return this `err`, defer is called with non-nil `err` and we close the connection, right? May be I am not quite following you

Comment (egonelbre) [reply internal/post/sender.go]: if the error is non-nil it doesn't mean the underlying connection is closed

File: internal/post/sender.go [hunk 3]
```diff
@@ -35,22 +36,27 @@ func (sender *SMTPSender) FromAddress() Address {
...
-	// close underlying connection
-	defer func() {
-		err = errs.Combine(err, client.Close())
-	}()
+
+	if err = sender.communicate(ctx, client, msg); err != nil {
+		return errs.Combine(err, client.Close())
+	}
+
+	return nil
```
Comment (egonelbre) [inline internal/post/sender.go]: keep the `client.Quit` here, that way it's more clear that it gets closed, also that way you can remove write data and keep it inside

Comment (rikysya) [reply internal/post/sender.go]: Then I need to remove `StartTls` and `Auth` out from there, so then I can possibly reuse communicate method on open connection, because calling `client.Quit` here makes me think that I can still use the `client` before quitting it. Another reason is that if `client.Quit` returns an`error` than `client.Close` should be called explicitly  

`writedata` makes sure that `writer` is closed immediately  after it is used, also it makes `error` more clear as `writer.Close` error will be combined only with `writer.Write` error
